identity-aware-proxy/IAPClient/IAPClient.cs

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2018 Google Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// [START generate_iap_request]
+
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Json;
+using Microsoft.IdentityModel.Tokens;
+using Newtonsoft.Json;
+
+namespace GoogleCloudSamples
+{
+    class IAPClient
+    {
+        public static string InvokeRequest(string iapClientId,
+            string credentialsPath, string uri)
+        {
+            // Read credentials from the credentials .json file.
+            Credentials credentials;
+            using (var fs = new FileStream(credentialsPath,
+                FileMode.Open, FileAccess.Read))
+            {
+                credentials = NewtonsoftJsonSerializer.Instance
+                    .Deserialize<Credentials>(fs);
+            }
+
+            // Generate a JWT signed with the service account's private key 
+            // containing a special "target_audience" claim.
+            var jwtBasedAccessToken =
+                CreateAccessToken(credentials.PrivateKey, iapClientId,
+                    credentials.ClientEmail);
+
+            var body = new Dictionary<string, string>
+            {
+                { "assertion", jwtBasedAccessToken },
+                { "grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"}
+            };
+
+            // Request an OIDC token for the Cloud IAP-secured client ID.
```

Comment:
Perhaps, worth extracting all the code that performs http post request to get the token into a separate method. It will make it easier to follow - people will understand the stages of the process better.

Reply:
If this code were a library, I'd totally agree.  But for samples, the clearest code is top-to bottom.